db/c.cc

```diff
@@ -414,60 +414,65 @@ struct rocksdb_flushjobinfo_t {
   const FlushJobInfo* rep;
 };
 
-struct rocksdb_event_listener_t : public EventListener {
+struct rocksdb_event_listener_inner_t : public EventListener {
   void* state_;
   void (*destructor_)(void*);
-  const char* (*name_)(void*);
-  void (*on_flush_completed_)(void*, rocksdb_t* db,
-                              const rocksdb_flushjobinfo_t* info);
-
-  rocksdb_event_listener_t() : EventListener() {}
+  void (*on_flush_completed_)(void*, const rocksdb_flushjobinfo_t* info);
 
-  ~rocksdb_event_listener_t() override { (*destructor_)(state_); }
+  rocksdb_event_listener_inner_t(void* state, void (*destructor)(void*))
+      : EventListener() {
+    state_ = state;
+    destructor_ = destructor;
+    on_flush_completed_ = nullptr;
+  }
 
-  const char* Name() const override {
-    if (name_ != nullptr) {
-      return (*name_)(state_);
-    } else {
-      return this->Name();
+  ~rocksdb_event_listener_inner_t() override {
+    if (destructor_ != nullptr) {
+      (*destructor_)(state_);
     }
   }
 
-  void OnFlushCompleted(DB* db, const FlushJobInfo& flush_job_info) override {
-    if (on_flush_completed_) {
+  void OnFlushCompleted(DB*, const FlushJobInfo& flush_job_info) override {
+    if (on_flush_completed_ != nullptr) {
       rocksdb_flushjobinfo_t info;
       info.rep = &flush_job_info;
-      (*on_flush_completed_)(state_, reinterpret_cast<rocksdb_t*>(db), &info);
+      (*on_flush_completed_)(state_, &info);
     }
   }
 };
 
-rocksdb_event_listener_t* rocksdb_event_listener_create(
-    void* state, void (*destructor_)(void*), const char* (*name)(void*)) {
-  rocksdb_event_listener_t* listener = new rocksdb_event_listener_t;
-  listener->state_ = state;
-  listener->destructor_ = destructor_;
-  listener->name_ = name;
+struct rocksdb_event_listener_t {
+  std::shared_ptr<rocksdb_event_listener_inner_t> inner_;
 
-  listener->on_flush_completed_ = nullptr;
+  rocksdb_event_listener_t(void* state, void (*destructor)(void*)) {
+    auto inner = new rocksdb_event_listener_inner_t(state, destructor);
+    inner_ = std::shared_ptr<rocksdb_event_listener_inner_t>(inner);
+  }
+};
 
-  return listener;
+// Once registered using rocksdb_options_add_event_listener(), the listener must outlive the corresponding DB.
+// Release with rocksdb_event_listener_destroy() when no longer needed.
```

Comment:
I'm not sure if this comment is necessary. I find it more confusing than helpful. Perhaps only focus on the fact that one need to call `rocksdb_event_listener_destroy()` after adding passing down the listener to `rocksdb_add_event_listener()` only?

Reply:
Also, checkout my comment https://github.com/restatedev/rocksdb/pull/7#discussion_r2065703613 which suggests to go back to your original direction to keep things consistent within the rocksdb c binding API.